csrc/optimization/remove_empty.cpp

```diff
@@ -0,0 +1,469 @@
+//!   a CatOp) then we replace the pad with `full(pad_value)`.
+//!   5. If empty TensorViews are input to an MmaOp and they are empty in
+//!   contracted axes, we replace with `full({m, n}, zeroVal())`.
+//!
+class EmptyTensorRemover : public DeadCodeRemover {
+ public:
+  EmptyTensorRemover(Fusion* fusion) : DeadCodeRemover(fusion) {}
+
+ protected:
+  using DeadCodeRemover::handle;
+
+  //! If tv is a fusion output, we check whether it is empty and if so, replace
+  //! it with full(). For non-outputs that are not inputs, we simply check that
+  //! the tensor is not provably empty.
+  void handle(TensorView* tv) final {
+    if (tv->isFusionInput()) {
+      // Skip inputs since they do not have a definition to redefine
+      return;
+    }
+
+    if (tv->isFusionOutput()) {
+      const auto rfactor =
+          TensorDomain::noReductions(tv->getMaybeRFactorDomain());
+      const auto empty_axes = emptyAxes(rfactor);
+      if (!empty_axes.empty()) {
+        std::vector<Val*> shape(rfactor.size());
+        std::transform(
+            rfactor.begin(), rfactor.end(), shape.begin(), [](IterDomain* id) {
+              return id->extent();
+            });
+        for (auto ax : empty_axes) {
+          shape[ax] = fusion()->zeroVal();
+        }
+        auto dtype = tv->getDataType().value();
+        auto new_tv = full(shape, fusion()->zeroVal(dtype), dtype);
+        replaceTV(tv, new_tv);
+      }
+    } else if (allUsesDead(tv)) {
+      // TensorViews that are not Fusion inputs or outputs and which have no
+      // uses are dead, so remove them and skip processing them and their
+      // definition.
+      markDeadAndMaybeRemove(tv);
+    } else {
+      TORCH_INTERNAL_ASSERT(
+          !isTVEmpty(tv),
+          "Found unexpected empty intermediate TensorView ",
+          tv->toString());
+    }
+  }
+
+  //! Gets a vector of extents for noReduction(tv->getMaybeRFactorDomain())
+  static std::vector<Val*> noReductionShape(TensorView* tv) {
+    std::vector<Val*> shape;
+    for (auto id : TensorDomain::noReductions(tv->getMaybeRFactorDomain())) {
+      shape.push_back(id->extent());
+    }
+    return shape;
+  }
+
+  //! A reduction over empty axes is equal to the initial value of the
+  //! reduction, as if the reduction were written as follows:
+  //!
+  //!   auto result = init_value;
+  //!   for (auto element : reduction_elements) {
+  //!     result = reduction_op(result, element);
+  //!   }
+  //!   return result;
+  //!
+  void handle(ReductionOp* rop) final {
+    auto in = rop->in()->as<TensorView>();
+    auto empty_input_axes =
+        emptyAxes(TensorDomain::noReductions(in->getMaybeRFactorDomain()));
+    if (empty_input_axes.empty()) {
+      // Input is not empty, handle like any other op
+      return;
+    }
+    auto out = rop->out()->as<TensorView>();
+    // The input is empty in some axes. Assert that they are all reduced
```

Comment:
In this case, this is just fine, but just FYI, the general method to match a producer and its consumer should be done with `PairwiseRootDomainMap`.